chinese_tutorial_rvest_httr.Rmd

```diff
@@ -0,0 +1,490 @@
+
+- `GET( )`
+
+`GET( )`函数使用的是`GET`请求方法。
+
+```R
+u <- "https://movie.douban.com/j/search_subjects?type=movie&tag=热门&page_limit=40&page_start=0"
+r <- GET(u,verbose())
+r$status_code
+```
+
+- `POST( )`
+
+在POST方法中，这三个部分：`status_line`,`headers`,`body`，都比较重要。
+
+我们对`status_line`中的`status_code`最感兴趣，因为它反映了我们的请求是否被接受，不被接受的话，又是因为什么而拒绝我们的请求。
+
+`POST()`函数向服务器以 POST 方式发起请求。我们使用`add_headers()`往请求中构造请求,在这里，主要讲解怎么在`POST( )`中构造`body`。
+
+```R
+url <- "http://httpbin.org/post"
+body <- list(a=1,b=2,c=3)
+r <- POST(url,body=body,encode="form",verbose())
+```
+
+**注释**：只有当`body`是**命名列表**时，我们才可以指定`encode`参数，并且`encode`参数的值又随`content-type`的值而有所不同。
+
+具体使用哪个函数发起请求，要依目标链接的请求方式而定。
+
+- `add_headers()`
+
+`add_headers()`可以在前面讲过的`GET()`、`POST()`请求中构造请求头。
+
+Q：为什么要构造请求头？
+
+A：反反爬策略之一。让你的请求成功通过，从而拿到你想要的数据。
+
+往请求标头里增加属性，我们使用函数`add_headers(name1=val1，name2=val2,...)`。
+
+```R
+u <- "https://movie.douban.com/j/search_subjects?type=movie&tag=热门&page_limit=40&page_start=0"
+
+headers <- c("User-Agent"="Mozilla/5.0 (Windows NT 10.0; WOW64; rv:51.0) Gecko/20100101 Firefox/51.0",
+             "X-Requested-With"="XMLHttpRequest",
+             "Cookie"='ll="108300"; bid=lVscia-_MWA; _pk_ref.100001.4cf6=%5B%22%22%2C%22%22%2C1499860968%2C%22https%3A%2F%2Fwww.baidu.com%2Fbaidu%3Fwd%3D%25E8%25B1%2586%25E7%2593%25A3%26tn%3Dmonline_4_dg%26ie%3Dutf-8%22%5D; _pk_id.100001.4cf6=c1f0706d81e95c0b.1488797655.52.1499860968.1499858546.; __utma=30149280.647508500.1488797657.1499856903.1499860968.52; __utmz=30149280.1499418495.32.17.utmcsr=douban.com|utmccn=(referral)|utmcmd=referral|utmcct=/; __utma=223695111.2135525097.1488797657.1499856903.1499860968.52; __utmz=223695111.1499860968.52.24.utmcsr=baidu|utmccn=(organic)|utmcmd=organic|utmctr=%E8%B1%86%E7%93%A3; _vwo_uuid_v2=E6655FD962C6D1103E3FF2A4F47436B5|9fe9945dedabcf10658c10bb20c79bd9; viewed="3283973"; gr_user_id=39d25c13-cc3d-4203-90b3-998f5d747d38; __yadk_uid=BYBkYyvVzLcrrN3BgQhaxvVdx4tJVLvt; ue="1329262214@qq.com"; __utmv=30149280.14318; ps=y; push_noty_num=0; push_doumail_num=0; ap=1; __utmc=30149280; __utmc=223695111; as="https://movie.douban.com/"; _pk_ses.100001.4cf6=*; __utmb=30149280.0.10.1499860968; __utmb=223695111.0.10.1499860968',
+             "Cache-Control"="max-age=0")
+
+r <- u %>% GET(add_headers(headers),verbose())
+
+bodyData <- r %>% content()
+```
+
+- `use_proxy()`
+
+`use_proxy(ip, port)`是用来为请求添加代理 ip 的。
+
+Q：为什么要添加代理 ip？
+
+A：反反爬策略之一。
+
+- `content()`
+
+`content()`用来获取响应的正文，即响应中的`body`部分。`body`中的内容可以是一个静态网页的 HTML 源码，也可以是在动态网页利用 AJAX 技术（异步加载）完成请求后，让后台返回的 JSON 格式数据。
+
+```R
+u <- "https://movie.douban.com/"
+
+# 返回静态网页 HTML 源码
+r <- u %>% GET(verbose())
+
+r %>% content() # 接下来可以用 rvest 包完成节点定位，获取数据的工作
+```
+
+```R
+## {xml_document}
+## <html lang="zh-cmn-Hans" class="">
+## [1] <head>\n<meta http-equiv="Content-Type" content="text/html; charset= ...
+## [2] <body>\n  \n    <script type="text/javascript">var _body_start = new ...
+```
+
+
+
+### 2.2 捕获错误机制
```

Comment:
Hashtag headers should not be followed by numbers since the hashtags will create numbered subheadings

Reply:
This problem has been solved, thank you.